rtl_wmbus.c

```diff
@@ -712,15 +716,52 @@ static void shift_freq_minus400(float *i, float *q, int fs_kHz)
     *q = cimagf(s);
 }
 
+
+static void shift_freq_plus_minus325(float *iplus, float *qplus,float *iminus, float *qminus, int fs_kHz)
+{
+    const int ft = 325;
+    static size_t n = 0;
+    const size_t n_max = fs_kHz/FREQ_STEP_KHZ;
+    #if 0
+    const float complex freq_shift = cosf(2.*M_PI*(ft*n)/fs_kHz) + sin(2.*M_PI*(ft*n)/fs_kHz) * _Complex_I;
+    n++;
+    #else
+    
+    const float x=crealf(LUT_FREQUENCY_TRANSLATION_PLUS[n]);
+    const float z=cimagf(LUT_FREQUENCY_TRANSLATION_PLUS[n]);
+    
+    n += ft/FREQ_STEP_KHZ;
+    #endif
+    //fprintf(stdout, "%ld, %ld: %f, %f\n", n, n_max, crealf(freq_shift), cimagf(freq_shift));
+    if (n >= n_max) n -= n_max;
+    float ix,iz,qx,qz; 
+    //(i+Jq)*(x+Jz) =ix-qz+ J(qx+iz) positive rotation
+    //(i+Jq)*(x-Jz) =ix+qz+ J(qx-iz) negative rotation
+    
+	ix=*iplus * x;
+	qx=*qplus * x;
+	iz=*iplus * z;
+	qz=*qplus * z;
+	//simetric negative shift 
+	*iminus=ix+qz; 
+	*qminus=qx-iz;
+	//simetric positive shift 
+	*iplus=ix-qz;
+	*qplus=qx+iz;	
```

Comment:
Did'not really check this. But something like this is usually more readable

 float complex freq_shift, s;

 freq_shift = LUT_FREQUENCY_TRANSLATION_PLUS[n];
 s = (*iplus) + (*qplus) * _Complex_I;
 s = s * freq_shift;
 *iplus = crealf(s);
 *qplus = cimagef(s);

 freq_shift = conjf(freq_shift);
 s = (*iminus) + (*qminus) * _Complex_I;
 s = s * freq_shift;
 *iminus = crealf(s);
 *qminus = cimagef(s);

But it's basically the same as before: two shifts in one function vs. two shifts slitted in two functions.

Reply:
Learnt to use gprof.
Verified with profiler, using s1_c1_samples.bin spends 2.14seconds to apply both frequency translations.
I'm saving costly float multiplications there.
Original spends with same sample file 4.84 seconds
The proposed readable using complex type only optimizes the number of function calls and runs in 3.95 seconds

The real bottleneck are FIR and IIR filters. 

optimiced gprof
index % time    self  children               called                   name
                2.14    0.00 111542272/111542272     shift_freq_plus_minus325 [11]
proposed readable using complex type
index % time    self  children               called                   name
              3.95    0.00 111542272/111542272     shift_freq_plus_minus325 [11]

original gprof
index % time    self  children               called                   name
                        2.47    0.00    111542272/111542272     shift_freq_minus400 [12]
                        2.37    0.00    111542272/111542272     shift_freq_250 [13]